File: src/chan/http/HttpSession.java
package chan.http;

import android.net.Uri;
import java.net.Proxy;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import okhttp3.Call;
import okhttp3.Headers;
import okhttp3.Response;

public final class HttpSession {
	final HttpHolder holder;
	final HttpClient client;
	HttpResponse response;

	final Proxy proxy;
	final boolean verifyCertificate;
	final boolean mayCheckFirewallBlock;
	final int delay;

	private final ArrayList<Uri> requestedUris = new ArrayList<>(2);
	Uri redirectedUri;

	int attempt;
	boolean forceGet;
	boolean executing;

	Call call;
	Response okResponse;
	Response deadResponse;
	HttpHolder.Callback callback;

	HttpSession(HttpHolder holder, HttpClient client, Uri uri, Proxy proxy,
			boolean verifyCertificate, boolean mayCheckFirewallBlock, int delay, int maxAttempts) {
		this.holder = holder;
		this.client = client;
		this.proxy = proxy;
		this.verifyCertificate = verifyCertificate;
		this.mayCheckFirewallBlock = mayCheckFirewallBlock;
		this.delay = delay;
		attempt = maxAttempts;
		requestedUris.add(uri);
	}

	void checkThread() {
		holder.checkThread();
	}

	void checkExecuting() {
		if (executing) {
			throw new IllegalStateException("Can't perform requests during active execute process");
		}
	}

	List<Uri> getRequestedUris() {
		checkThread();
		return requestedUris;
	}

	Uri getCurrentRequestedUri() {
		checkThread();
		return requestedUris.get(requestedUris.size() - 1);
	}

	void setNextRequestedUri(Uri uri) {
		checkThread();
		requestedUris.add(uri);
	}

	private void setCallInternal(Call call, HttpHolder.Callback callback)
			throws HttpClient.InterruptedHttpException {
		checkThread();
		this.call = call;
		this.callback = callback;
		redirectedUri = null;
		if (holder.isInterrupted()) {
			this.call = null;
			this.callback = null;
			throw new HttpClient.InterruptedHttpException();
		}
		if (call != null) {
			client.onConnect(holder.chan, call, delay);
		}
	}

	void setCall(Call call) throws HttpClient.InterruptedHttpException {
		checkThread();
		setCallInternal(call, null);
	}

	void setCallback(HttpHolder.Callback callback) throws HttpClient.InterruptedHttpException {
		checkThread();
		setCallInternal(null, callback);
	}

	boolean nextAttempt() {
		checkThread();
		return attempt-- > 0;
	}

	void disconnectAndClear() {
		checkThread();
		Call call = this.call;
		this.call = null;
		Response okResponse = this.okResponse;
		this.okResponse = null;
		HttpHolder.Callback callback = this.callback;
		this.callback = null;
		if (response != null) {
			// HttpResponse will call disconnectAndClear if the response is still active
			response.cleanupAndDisconnect();
		}
		if (okResponse != null) {
			okResponse.close();
			deadResponse = okResponse;
		}
		if (call != null) {
			call.cancel();
			client.onDisconnect(call);
		}
		if (callback != null) {
			callback.onDisconnectRequested();
		}
	}

	void checkResponseCode() throws HttpException {
		checkThread();
		int responseCode = getResponseCode();
		boolean success = responseCode >= 200 && responseCode <= 303
				|| responseCode == HttpClient.HTTP_TEMPORARY_REDIRECT;
		if (!success) {
			String message = HttpClient.transformResponseMessage(getResponseMessage());
			disconnectAndClear();
			throw new HttpException(responseCode, message);
		}
	}

	private Response getResponseForHeaders() {
		checkThread();
		Response response = okResponse;
		if (response == null) {
			response = deadResponse;
		}
		return response;
	}

	int getResponseCode() {
		Response response = getResponseForHeaders();
		return response != null ? response.code() : -1;
	}

	String getResponseMessage() {
		Response response = getResponseForHeaders();
		if (response != null) {
			String message = response.message();
			return message.isEmpty() ? null : message;
		}
		return null;
	}

	Map<String, List<String>> getHeaderFields() {
		Response response = getResponseForHeaders();
		if (response == null) {
			return Collections.emptyMap();
		}
		Headers headers = response.headers();
		LinkedHashMap<String, List<String>> map = new LinkedHashMap<>();
		for (int i = 0; i < headers.size(); i++) {
			String name = headers.name(i);
			List<String> values = map.get(name);
			if (values == null) {
				values = new ArrayList<>(1);
				map.put(name, values);
			}
			values.add(headers.value(i));
		}
		return map;
	}

	String getCookieValue(String name) {
		Map<String, List<String>> headers = getHeaderFields();
		List<String> cookies = headers.get("Set-Cookie");
		if (cookies == null) {
			cookies = headers.get("set-cookie");
		}
		if (cookies != null) {
			String start = name + "=";
			for (String cookie : cookies) {
				if (cookie.startsWith(start)) {
					int startIndex = start.length();
					int endIndex = cookie.indexOf(';');
					if (endIndex >= 0) {
						return cookie.substring(startIndex, endIndex);
					} else {
						return cookie.substring(startIndex);
					}
				}
			}
		}
		return null;
	}

	long getLength() {
		Response response = getResponseForHeaders();
		if (response != null && HttpClient.Encoding.get(response.headers()) == HttpClient.Encoding.IDENTITY) {
			String contentLength = response.header("Content-Length");
			if (contentLength != null) {
				try {
					return Long.parseLong(contentLength);
				} catch (NumberFormatException e) {
					return -1;
				}
			}
		}
		return -1;
	}
}

File: src/chan/http/HttpSession.kt
package chan.http

import android.net.Uri
import chan.http.HttpClient.InterruptedHttpException
import chan.util.CommonUtils.equals
import okhttp3.Call
import okhttp3.Response

class HttpSession internal constructor(
    val holder: HttpHolder,
    val client: HttpClient,
    uri: Uri?,
    val proxyData: HttpClient.ProxyData?,
    val verifyCertificate: Boolean,
    val mayCheckFirewallBlock: Boolean,
    val delay: Int,
    var attempt: Int,
) {
    var response: HttpResponse? = null

    private val requestedUris = ArrayList<Uri?>(2)
    var redirectedUri: Uri? = null

    var forceGet: Boolean = false
    var executing: Boolean = false

    var currentCall: Call? = null
    var okResponse: Response? = null
    var deadResponse: Response? = null
    var currentCallback: HttpHolder.Callback? = null

    init {
        requestedUris.add(uri)
    }

    fun checkThread() {
        holder.checkThread()
    }

    fun checkExecuting() {
        check(!executing) { "Can't perform requests during active execute process" }
    }

    fun getRequestedUris(): MutableList<Uri?> {
        checkThread()
        return requestedUris
    }

    val currentRequestedUri: Uri?
        get() {
            checkThread()
            return requestedUris[requestedUris.size - 1]
        }

    fun setNextRequestedUri(uri: Uri?) {
        checkThread()
        requestedUris.add(uri)
    }

    @Throws(InterruptedHttpException::class)
    private fun setCallInternal(
        call: Call?,
        callback: HttpHolder.Callback?,
    ) {
        checkThread()
        this.currentCall = call
        this.currentCallback = callback
        redirectedUri = null
        if (holder.isInterrupted) {
            this.currentCall = null
            this.currentCallback = null
            throw InterruptedHttpException()
        }
        if (call != null) {
            client.onConnect(holder.chan!!, call, delay)
        }
    }

    @Throws(InterruptedHttpException::class)
    fun setCall(call: Call?) {
        checkThread()
        setCallInternal(call, null)
    }

    @Throws(InterruptedHttpException::class)
    fun setCallback(callback: HttpHolder.Callback?) {
        checkThread()
        setCallInternal(null, callback)
    }

    fun nextAttempt(): Boolean {
        checkThread()
        return attempt-- > 0
    }

    fun disconnectAndClear() {
        checkThread()
        val call = this.currentCall
        this.currentCall = null
        val okResponse = this.okResponse
        this.okResponse = null
        val callback = this.currentCallback
        this.currentCallback = null
        // HttpResponse will call disconnectAndClear if the response is still active
        response?.cleanupAndDisconnect()
        if (okResponse != null) {
            okResponse.close()
            deadResponse = okResponse
        }
        if (call != null) {
            call.cancel()
            client.onDisconnect(call)
        }
        if (callback != null) {
            callback.onDisconnectRequested()
        }
    }

    @Throws(HttpException::class)
    fun checkResponseCode() {
        checkThread()
        val responseCode = this.responseCode
        val success =
            (responseCode >= 200 && responseCode <= 303) ||
                responseCode == HttpClient.Companion.HTTP_TEMPORARY_REDIRECT
        if (!success) {
            val message: String? =
                HttpClient.Companion.transformResponseMessage(
                    this.responseMessage,
                )
            disconnectAndClear()
            throw HttpException(responseCode, message)
        }
    }

    private val responseForHeaders: Response?
        get() {
            checkThread()
            var response = okResponse
            if (response == null) {
                response = deadResponse
            }
            return response
        }

    val responseCode: Int
        get() {
            val response = this.responseForHeaders
            return if (response != null) response.code else -1
        }

    val responseMessage: String?
        get() {
            val response = this.responseForHeaders
            if (response != null) {
                val message = response.message
                return if (message.isEmpty()) null else message
            }
            return null
        }

    /**
     * Keyed without regard to case, because the case a header arrives in is the protocol's choice,
     * not the server's: HTTP/2 lowercases every name, so a server that sent `CF-RAY` over HTTP/1.1
     * sends `cf-ray` here. Everything reading a header by its familiar spelling — the firewall
     * resolvers, the extensions — would otherwise silently find nothing on exactly the hosts modern
     * enough to negotiate HTTP/2.
     */
    val headerFields: MutableMap<String?, MutableList<String>?>
        get() {
            val response = this.responseForHeaders ?: return sortedByHeaderName()
            val headers = response.headers
            val map = sortedByHeaderName()
            for (i in 0..<headers.size) {
                val name = headers.name(i)
                var values = map[name]
                if (values == null) {
                    values = ArrayList(1)
                    map[name] = values
                }
                values.add(headers.value(i))
            }
            return map
        }

    fun getCookieValue(name: String?): String? {
        val cookies = this.headerFields["Set-Cookie"]
        if (cookies != null) {
            val start = name + "="
            for (cookie in cookies) {
                if (cookie.startsWith(start)) {
                    val startIndex = start.length
                    val endIndex = cookie.indexOf(';')
                    if (endIndex >= 0) {
                        return cookie.substring(startIndex, endIndex)
                    } else {
                        return cookie.substring(startIndex)
                    }
                }
            }
        }
        return null
    }

    val length: Long
        get() {
            val response = this.responseForHeaders
            if (response != null && HttpClient.Encoding.Companion.get(response.headers) == HttpClient.Encoding.IDENTITY) {
                val contentLength = response.header("Content-Length")
                if (contentLength != null) {
                    try {
                        return contentLength.toLong()
                    } catch (e: NumberFormatException) {
                        return -1
                    }
                }
            }
            return -1
        }

    companion object {
        /**
         * The map [headerFields] hands out. A null name never comes off the wire, but the map is
         * `@Public` with a nullable key type, so the order tolerates one rather than throwing where
         * the old [LinkedHashMap] would have returned null.
         */
        private val HEADER_NAME_ORDER =
            Comparator<String?> { first, second ->
                when {
                    first == null && second == null -> 0
                    first == null -> -1
                    second == null -> 1
                    else -> String.CASE_INSENSITIVE_ORDER.compare(first, second)
                }
            }

        private fun sortedByHeaderName(): MutableMap<String?, MutableList<String>?> = java.util.TreeMap(HEADER_NAME_ORDER)
    }
}
